materialize-bigquery/transactor.go

```diff
@@ -2,287 +2,271 @@ package main
+				}
 
-			// Clean up the temporary files when store complete (or we error out).
-			defer func(defb *binding) {
-				if err := defb.store.mergeFile.Delete(ctx); err != nil {
-					log.Errorf("could not delete store mergefile: %v", err)
+				// Load the document by sending it back into Flow.
+				if err = loaded(idx, fd.Body); err != nil {
+					return fmt.Errorf("load row loaded: %w", err)
 				}
-				defb.store.mergeFile = nil
-			}(b) // b is part of for loop, make sure you're referencing the right b.
-		}
+			}
+
+			return nil
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		return fmt.Errorf("one of the load call in the group failed: %w", err)
 	}
 
-	// Complete the transaction and return the appropriate error.
-	subqueries = append(subqueries, `
-	COMMIT TRANSACTION;
+	return nil
+}
 
-	SELECT '' AS error;
+func (t *transactor) Prepare(ctx context.Context, _ pm.TransactionRequest_Prepare) (pf.DriverCheckpoint, error) {
+	t.checkpoint = NewBigQueryCheckPoint()
 
-	EXCEPTION WHEN ERROR THEN
-		ROLLBACK TRANSACTION;
-		SELECT @@error.message AS error;
-	END;
-	`)
+	for i, binding := range t.bindings {
+		binding.SetWriter(
+			ctx,
+			NewWriter(ctx, binding, fmt.Sprintf("%s/%s", t.config.BucketPath, randomString())),
+		)
 
-	// Build the bigquery query of the combined subqueries.
-	query := t.ep.newQuery(strings.Join(subqueries, "\n"), args...)
-	query.TableDefinitions = edcTableDefs // Tell the query where to get the external references in gcs.
+		t.checkpoint.Bindings = append(t.checkpoint.Bindings, DriverCheckPointBinding{
+			FilePath:     binding.FilePath(),
+			Query:        binding.CreateSQL,
+			BindingIndex: i,
+		})
+	}
 
-	// This returns a single row with the error status of the query.
-	job, err := t.ep.runQuery(ctx, query)
+	jsn, err := json.Marshal(t.checkpoint)
 	if err != nil {
-		return fmt.Errorf("load query: %w", err)
+		return pf.DriverCheckpoint{}, fmt.Errorf("creating checkpoint json: %w", err)
 	}
 
-	// Fetch te query status from the transaction
-	var queryStatus struct {
-		Error string `bigquery:"error"`
-	}
+	return pf.DriverCheckpoint{
+		DriverCheckpointJson: jsn,
+	}, nil
+}
 
-	if err := t.ep.fetchOne(ctx, job, &queryStatus); err != nil {
-		return fmt.Errorf("fetch one: %w", err)
+func (t *transactor) Store(it *pm.StoreIterator) error {
+	for it.Next() {
+		binding := t.bindings[it.Binding]
+		binding.Store(binding.GenerateDocument(it.Key, it.Values, it.RawJSON))
 	}
+	return nil
+}
 
-	if queryStatus.Error != "" {
-		return fmt.Errorf("merge error: %s", queryStatus.Error)
+func (t *transactor) Commit(ctx context.Context) error {
+	for _, binding := range t.bindings {
+		err := binding.Commit(ctx)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
-func (t *transactor) Acknowledge(context.Context) error {
+func (t *transactor) Acknowledge(ctx context.Context) error {
+	for _, cpBinding := range t.checkpoint.Bindings {
+		binding := t.bindings[cpBinding.BindingIndex]
```

Comment:
Per my comment above, we can't do this here. We should only depend on `t.checkpoint` alone for this. `t.binding` is based on the new transaction's bindings, not the previous one. Our only point of reference is checkpoint in order to process the previous transaction.